src/tags/tests/test_models.py

```diff
@@ -0,0 +1,247 @@
+            badge.tags.add("TAG")
+
+        self.assertEqual(total_xp_by_tags(self.user, ["TAG"]), sum(xp_list))
+
+    def test_multiple_separate_tags_quests_badges(self):
+        """ 
+            See if correct xp is displayed using only multiple tags added to Quest and Badge objects
+            only 1 unique tag per quest and badge
+        """ 
+        tag_list = [f"tag-{i}" for i in range(5 * 2)]
+        xp_list = [self.random.randint(0, 100) for i in range(5 * len(tag_list))]
+        xp_iter = itertools.cycle(xp_list)
+
+        for tag_el in tag_list:
+            for i in range(len(xp_list) // len(tag_list)):
+
+                if i % 2 == 0:
+                    quest, _ = self.create_quest(next(xp_iter))
+                    quest.tags.add(tag_el)
+                else:
+                    badge, _ = self.create_badge(next(xp_iter))
+                    badge.tags.add(tag_el)
+
+        self.assertEqual(total_xp_by_tags(self.user, tag_list), sum(xp_list))
+
+    def test_multiple_crossing_tags_quest_badges(self):
+        """ 
+            See if correct xp is displayed using only multiple tags added to Quest and Badge objects
+            each Quest and Badge object should have multiple tags
+        """ 
+        tag_list = [(f"tag0-{i}", f"tag1-{i}") for i in range(5 * 2)]
+        xp_list = [self.random.randint(0, 100) for i in range(5 * len(tag_list))]
+        xp_iter = itertools.cycle(xp_list)
+
+        for tag_set in tag_list:
+            for i in range(len(xp_list) // len(tag_list)):
+                
+                if i % 2 == 0:
+                    quest, _ = self.create_quest(next(xp_iter))
+                    quest.tags.add(tag_set[0])
+                    quest.tags.add(tag_set[1])
+                else:
+                    badge, _ = self.create_badge(next(xp_iter))
+                    badge.tags.add(tag_set[0])
+                    badge.tags.add(tag_set[1])
+
+        unpacked_tag_list = [tag_set[index] for tag_set in tag_list for index in range(len(tag_set))]
+        self.assertEqual(total_xp_by_tags(self.user, unpacked_tag_list), sum(xp_list))
+
+    def test_one_tag_ordinal_quests_badges(self):
+        """ 
+             See if correct xp is displayed using only one tag + quests and badges with ordinal > 1 
+        """ 
+        ordinal_list = [self.random.randint(1, 5) for i in range(10 * 2)]
+        xp_list = [self.random.randint(0, 100) for i in range(len(ordinal_list))]
+        ordinal_iter = itertools.cycle(ordinal_list)
+        xp_iter = itertools.cycle(xp_list)
+        
+        for i in range(len(xp_list) // 2):
+            quest, qs = self.create_quest(next(xp_iter), next(ordinal_iter))
+            quest.tags.add("TAG")
+
+            badge, ba = self.create_badge(next(xp_iter), next(ordinal_iter))
+            badge.tags.add("TAG")
+
+        self.assertEqual(total_xp_by_tags(self.user, ["TAG"]), sum(xp_list[i] * ordinal_list[i] for i in range(len(ordinal_list))))
```

Comment:
As I mentioned above, I'm not sure whether this code is all-time XP or current-semester-only xp?  This will depend on the defaults of those methods you are using (it should say in their docstrings, or be obvious from their code).  Both are useful, however most important is the xp value earned from a tag in the current semester only.

In many model manager methods you'll see a boolean parameter `current_sem_only` or something like that, perhaps you could do the same thing here?

Regardless, we will need a test to verify that the Badge and QuestSubmission/Quest queries are doing the same thing, and which one it is: all semesters or current semester.  We definitly don't want, for example, Quests to be all-semesters and Badges to be current semester only, or vice-versa.

Reply:
Fixed
Added a list of 50 ints that im calling/slicing instead of calling ``self.random.randint()``

Added an extra arg for badge filter() ``semester=SiteConfig.get().active_semester`` + added test for that